[UNIT] smolagents framework - release 24/2/25 (#102)
huggingface/agents-course: units/en/unit2/smolagents/tools.md

# Tools  

> A tool is essentially a function that an LLM can use within an agentic system. To interact with it, the LLM requires an API with the following components:  

Reviewer: ```suggestion
As we learnt in unit 1, agents use tools to perform actions. In `smolagents` tools are treated as functions that an LLM can use within an agentic system. To interact with a tool, the LLM requires an **interface description** with the following components:  
```

- **Name**  
- **Tool description**  
- **Input types and their descriptions**  
- **Output type**  

> Rather than using a regular Python function, a tool should be implemented as a class. Essentially, it wraps a function with metadata that helps the LLM understand how to use it effectively.  

Reviewer: ```suggestion
For complex tools, we can implement a class instead of a Python function. The class wraps the function with metadata that helps the LLM understand how to use it effectively.  
```

- `output_type`: Specifies the expected output type.  
- `forward`: The method containing the inference logic to execute.

Below, we can see an example of a tool built using `Tool` and to integrate it within a `CodeAgent`.

```python
from smolagents import Tool

class HFModelDownloadsTool(Tool):
    name = "model_download_counter"
    description = """
    This is a tool that returns the most downloaded model of a given task on the Hugging Face Hub.
    It returns the name of the checkpoint."""
    inputs = {
        "task": {
            "type": "string",
            "description": "the task category (such as text-classification, depth-estimation, etc)",
        }
    }
    output_type = "string"

    def forward(self, task: str):
        from huggingface_hub import list_models

        model = next(iter(list_models(filter=task, sort="downloads", direction=-1)))
        return model.id

model_downloads_tool = HFModelDownloadsTool()

agent = CodeAgent(tools=[model_downloads_tool], model=HfApiModel())

agent.run(
    "Can you give me the name of the model that has the most downloads in the 'text-to-video' task on the Hugging Face Hub?"
)
```

> ### `@tool` Decorator  

Reviewer: I would this section before the class implementation.

> The `@tool` decorator is the recommended way to define simple tools. Using this approach, we define a function with:  

Reviewer: ```suggestion
The `@tool` decorator is the recommended way to define simple tools. Under the hood, smolagents will parse basic information about the function from Python. So if you name you function clearly and a good docstring, it will be easier for the LLM to use. Using this approach, we define a function with:  
```

agent.run(
    "Can you give me the name of the model that has the most downloads in the 'text-to-video' task on the Hugging Face Hub?"
)
```

## Default Toolbox  

`smolagents` comes with a set of pre-built tools that can be directly injected into your agent. The [default toolbox](https://huggingface.co/docs/smolagents/guided_tour?build-a-tool=Decorate+a+function+with+%40tool#default-toolbox) includes:  

- **PythonInterpreterTool**  
- **FinalAnswerTool**  
- **UserInputTool**  
- **DuckDuckGoSearchTool**  
- **GoogleSearchTool**  
- **VisitWebpageTool**  

## Sharing and Importing Tools

One of the most powerful features of **smolagents** is the ability to share your custom tools to the Hub, as well as load tools shared by the community. This includes integrating **HF Spaces** or **LangChain tools**. Below are examples showcasing each of these functionalities:

### Sharing a Tool to the Hub

To share your custom tool, you can upload it to your Hugging Face account using the `push_to_hub()` method:

```python
model_downloads_tool.push_to_hub("{your_username}/hf-model-downloads", token="<YOUR_HUGGINGFACEHUB_API_TOKEN>")
```

> ### Importing a Tool from the Hub

Reviewer 2: This part looks quite similar to [this doc page](https://huggingface.co/docs/smolagents/tutorials/tools).
Maybe it could be integrated more in telling a story?

You can import tools developed by other users by utilizing the load_tool() function:

```python
from smolagents import load_tool, CodeAgent

model_download_tool = load_tool(
    "{your_username}/hf-model-downloads", # m-ric/text-to-image
    trust_remote_code=True
)
```

### Importing a Hugging Face Space as a Tool

> You can also import a HF Space as a tool using `Tool.from_space()`. This opens up many possibilities for integration. The functionality uses the `gradio_client` under the hood, so make sure to install it via `pip` if you don't have it already:

Reviewer: ```suggestion
You can also import a HF Space as a tool using `Tool.from_space()`. This opens up possibilities for integrating with thousands of spaces from the community for tasks from image generation to data analysis. The tool will connect with the spaces Gradio backend using the `gradio_client`, so make sure to install it via `pip` if you don't have it already:
```
